genai-perf/tests/test_llm_goodput_calculator.py

```diff
@@ -0,0 +1,146 @@
+            metric=self.TEST_METRIC,
+            benchmark_duration=self.TEST_BENCHMARK_DURATION,
+        )
+
+        assert gc.goodput is None  # before computing
+        gc.compute()
+        assert gc.goodput == [0.1]  # after computing
+
+    def test_get_slo_name(self) -> None:
+        gc = LLMGoodputCalculator(
+            goodput_constraints=self.TEST_GOODPUT_CONSTRAINTS,
+            metric=self.TEST_METRIC,
+            benchmark_duration=self.TEST_BENCHMARK_DURATION,
+        )
+
+        assert gc.get_slo_name("request_latency") == "request_latencies"
+        assert gc.get_slo_name("time_to_first_token") == "time_to_first_tokens"
+        assert gc.get_slo_name("inter_token_latency") == "inter_token_latencies"
+        assert gc.get_slo_name("output_token_throughput_per_request") == (
+            "output_token_throughputs_per_request"
+        )
+        with pytest.raises(KeyError):
+            gc.get_slo_name("hello1234")
+
+    def test_compute(self) -> None:
+        """
+        Goodput constraints for experiment 1 and 2:
+        * time_to_first_token: 2.5e-6 ms
+        * inter_token_latency: 2.5e-6 ms
+        * output_token_throughput_per_request: 0.5e9 s
+
+        Benchmark duration for experiment 1 and 2: 10 s
+
+        LLMMetrics
+        * time to first tokens
+            - experiment 1: [3 - 1, 4 - 2] = [2, 2]
+            - experiment 2: [7 - 5, 6 - 3] = [2, 3]
+        * inter token latencies
+            - experiment 1: [((8 - 1) - 2)/(3 - 1), ((11 - 2) - 2)/(6 - 1)]
+                          : [2.5, 1.4]
+                          : [2, 1]  # rounded
+            - experiment 2: [((18 - 5) - 2)/(4 - 1), ((11 - 3) - 3)/(6 - 1)]
+                          : [11/3, 1]
+                          : [4, 1]  # rounded
+        * output token throughputs per request
+            - experiment 1: [3/(8 - 1), 6/(11 - 2)] = [3/7, 6/9]
+            - experiment 2: [4/(18 - 5), 6/(11 - 3)] = [4/13, 6/8]
+
+        Request good counts according to constraints:
+            - experiment 1: 1
+            - experiment 2: 0
+
+        Request goodputs
+            - experiment 1: [1 / 10] = [0.1]
+            - experiment 2: [0 / 10] = [0]
+        """
+        test_goodput_constraints = {
+            "time_to_first_token": 2.5e-6,  # ms
+            "inter_token_latency": 1.5e-6,  # ms
+            "output_token_throughput_per_request": 0.5e9,  # s
+        }
+
+        # experiment 1
+        test_llm_metrics_1 = LLMMetrics(
+            time_to_first_tokens=[2, 2],
+            inter_token_latencies=[2, 1],
+            output_token_throughputs_per_request=[3 / ns_to_sec(7), 6 / ns_to_sec(9)],
+        )
+
+        gc_1 = LLMGoodputCalculator(
+            goodput_constraints=test_goodput_constraints,
+            metric=test_llm_metrics_1,
+            benchmark_duration=self.TEST_BENCHMARK_DURATION,
+        )
+        assert gc_1.goodput is None  # before computing
```

Comment:
These comments seem like they can be removed. The code is clear and self-commenting here.